godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,818 @@
+    unsafe {
+        sys::interface_fn!(script_instance_create)(
+            instance_ptr,
+            data_ptr as sys::GDExtensionScriptInstanceDataPtr,
+        )
+    }
+}
+
+mod script_instance_info {
+    use std::any::type_name;
+    use std::cell::{BorrowError, Ref, RefMut};
+    use std::ffi::c_void;
+    use std::mem::ManuallyDrop;
+
+    use crate::builtin::meta::GodotType;
+    use crate::builtin::{GString, StringName, Variant};
+    use crate::engine::{Object, ScriptLanguage};
+    use crate::obj::Gd;
+    use crate::private::handle_panic;
+    use crate::sys::{self, GodotFfi};
+
+    use super::{ScriptInstance, ScriptInstanceData};
+
+    fn borrow_instance_mut<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> RefMut<'a, T> {
+        unsafe { (*instance).inner.borrow_mut() }
+    }
+
+    fn borrow_instance<'a, T: ScriptInstance>(instance: *mut ScriptInstanceData<T>) -> Ref<'a, T> {
+        unsafe { (*instance).inner.borrow() }
+    }
+
+    fn try_borrow_instance<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> Result<Ref<'a, T>, BorrowError> {
+        unsafe { (*instance).inner.try_borrow() }
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a vallid instance pointer.
+    fn instance_data_as_script_instance<T: ScriptInstance>(
+        p_instance: sys::GDExtensionScriptInstanceDataPtr,
+    ) -> *mut ScriptInstanceData<T> {
+        p_instance as *mut ScriptInstanceData<T>
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a valid const string name pointer.
+    unsafe fn transfer_string_name_from_godot(
+        p_name: sys::GDExtensionConstStringNamePtr,
+    ) -> StringName {
+        StringName::from_arg_ptr(
+            p_name as sys::GDExtensionTypePtr,
+            sys::PtrcallType::Standard,
+        )
+    }
+
+    fn transfer_bool_to_godot(value: bool) -> u8 {
+        value as u8
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a valid variant pointer the return value can be moved into.
+    unsafe fn transfer_variant_to_godot(variant: Variant, return_ptr: sys::GDExtensionVariantPtr) {
+        variant.move_return_ptr(
+            return_ptr as sys::GDExtensionTypePtr,
+            sys::PtrcallType::Standard,
+        )
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid `*mut u32` pointer for the list length.
+    /// - The list pointer is guaranteed to point to a list that has an equal length and capacity.
+    unsafe fn transfer_ptr_list_to_godot<T>(ptr_list: Vec<T>, list_length: *mut u32) -> *const T {
+        *list_length = ptr_list.len() as u32;
+
+        let ptr = Box::into_raw(ptr_list.into_boxed_slice());
+
+        (*ptr).as_ptr()
+    }
+
+    fn transfer_script_to_godot(script: Gd<crate::engine::Script>) -> sys::GDExtensionObjectPtr {
+        script.obj_sys()
+    }
```

Comment:
This will erroneously decrement the refcount. As the `script` gets dropped but you return a pointer to `script` anyway. So if the refcount is 1 then the script will be freed and this method returns a null-pointer. If you can't avoid this pattern, then you must ensure that `script` is not dropped.

Reply:
I am aware, there is another thread about this above, so I will close this one, thank you.